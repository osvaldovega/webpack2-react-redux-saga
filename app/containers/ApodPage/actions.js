/*
 * APOD Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_APOD,
  LOAD_APOD_SUCCESS,
  LOAD_APOD_ERROR,
} from './constants';

/**
 * Load the APOD data, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_APOD
 */
export function loadApod() {
  return {
    type: LOAD_APOD,
  };
}

/**
 * Dispatched when the data are loaded by the request saga
 *
 * @param  {object} Object that has the APOD data
 *
 * @return {object} An action object with a type of LOAD_APOD_SUCCESS passing the repos
 */
export function loadApodSuccess(data) {
  return {
    type: LOAD_APOD_SUCCESS,
    payload: data,
  };
}

/**
 * Dispatched when loading the data from NASA fails
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of LOAD_APOD_ERROR passing the error
 */
export function loadApodError(error) {
  return {
    type: LOAD_APOD_ERROR,
    error,
  };
}
