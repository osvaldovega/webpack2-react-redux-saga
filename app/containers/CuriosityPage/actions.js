/*
 * CURIOSITY Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_CURIOSITY,
  LOAD_CURIOSITY_SUCCESS,
  LOAD_CURIOSITY_ERROR,
} from './constants';

/**
 * Load the CURIOSITY data, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_CURIOSITY
 */
export function loadCuriosity() {
  return {
    type: LOAD_CURIOSITY,
  };
}

/**
 * Dispatched when the data are loaded by the request saga
 *
 * @param  {object} Object that has the CURIOSITY data
 *
 * @return {object} An action object with a type of LOAD_CURIOSITY_SUCCESS passing the repos
 */
export function loadCuriositySuccess(data) {
  return {
    type: LOAD_CURIOSITY_SUCCESS,
    payload: data,
  };
}

/**
 * Dispatched when loading the data from NASA fails
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of LOAD_CURIOSITY_ERROR passing the error
 */
export function loadCuriosityError(error) {
  return {
    type: LOAD_CURIOSITY_ERROR,
    error,
  };
}
