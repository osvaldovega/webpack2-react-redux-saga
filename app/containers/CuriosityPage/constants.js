/*
 * CuriosityConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const LOAD_CURIOSITY = 'universe/curiosity/LOAD_CURIOSITY';
export const LOAD_CURIOSITY_SUCCESS = 'universe/curiosity/LOAD_CURIOSITY_SUCCESS';
export const LOAD_CURIOSITY_ERROR = 'universe/curiosity/LOAD_CURIOSITY_ERROR';
