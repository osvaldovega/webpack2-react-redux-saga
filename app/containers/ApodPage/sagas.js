import { take, call, put, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { LOAD_APOD } from './constants';
import { loadApodSuccess, loadApodError } from './actions';
import request from '../../utils/request';
import nasa from '../../utils/endpointNASA.json';

/**
 * Build NASA URL
 */
function generateURL(partialURL) {
  return `${nasa.base_url}${partialURL}${nasa.key}`;
}

/**
 * NASA APOD request/response handler
 */
export function* getApod() {
  const requestURL = generateURL(nasa.apod_url);

  try {
    // Call our request helper (see 'utils/request')
    const data = yield call(request, requestURL);
    yield put(loadApodSuccess(data));
  } catch (err) {
    yield put(loadApodError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* apodData() {
  // Watches for LOAD_REPOS actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  const watcher = yield takeLatest(LOAD_APOD, getApod);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  apodData,
];
