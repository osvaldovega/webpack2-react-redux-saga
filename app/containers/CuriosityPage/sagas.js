import { take, call, put, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { LOAD_CURIOSITY } from './constants';
import { loadCuriositySuccess, loadCuriosityError } from './actions';
import request from '../../utils/request';
import nasa from '../../utils/endpointNASA.json';

/**
 * Build NASA URL
 */
function generateURL(partialURL) {
  return `${nasa.base_url}${partialURL}${nasa.key}`;
}

/**
 * NASA APOD request/response handler
 */
export function* getCuriosity() {
  const requestURL = generateURL(nasa.rovers_url_1);

  try {
    // Call our request helper (see 'utils/request')
    const data = yield call(request, requestURL);
    yield put(loadCuriositySuccess(data));
  } catch (err) {
    yield put(loadCuriosityError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* CuriosityData() {
  // Watches for LOAD_REPOS actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  const watcher = yield takeLatest(LOAD_CURIOSITY, getCuriosity);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  CuriosityData,
];
