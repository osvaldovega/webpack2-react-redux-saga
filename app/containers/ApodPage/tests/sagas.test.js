/**
 * Tests for HomePage sagas
 */

import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';

import { LOCATION_CHANGE } from 'react-router-redux';

import { LOAD_APOD } from '../constants';
import { loadApodSuccess, loadApodError } from '../actions';

import { getApod, apodData } from '../sagas';


/* eslint-disable redux-saga/yield-effects */
describe('getApod Saga', () => {
  let getApodGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getApodGenerator = getApod();

    const selectDescriptor = getApodGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should dispatch the loadApodSuccess action if it requests the data successfully', () => {
    const response = [{
      name: 'First repo',
    }];
    const putDescriptor = getApodGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(loadApodSuccess(response)));
  });

  it('should call the loadApodError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getApodGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(loadApodError(response)));
  });
});

describe('apodDataSaga Saga', () => {
  const apodDataSaga = apodData();
  const mockedTask = createMockTask();

  it('should start task to watch for LOAD_APOD action', () => {
    const takeLatestDescriptor = apodDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_APOD, getApod));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = apodDataSaga.next(mockedTask).value;
    expect(takeDescriptor).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = apodDataSaga.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockedTask));
  });
});
