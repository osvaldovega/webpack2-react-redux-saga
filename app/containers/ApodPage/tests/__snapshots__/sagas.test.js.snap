exports[`getApod Saga should call the loadApodError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.nasa.gov/planetary/apod?api_key=0LDuEGPf0Zbh1hVhIkPekDY6nOUy5aOmgKB6wAYS",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getApod Saga should dispatch the loadApodSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.nasa.gov/planetary/apod?api_key=0LDuEGPf0Zbh1hVhIkPekDY6nOUy5aOmgKB6wAYS",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
